C/plugins/storage/postgres/connection.cpp

```diff
@@ -3312,3 +3312,422 @@ const char *p;
 		return true;
 	return false;
 }
+
+bool Connection::findSchemaFromDB(std::string name, std::string &resultSet)
+{
+
+	SQLBuffer sql;
+        try
+        {
+		sql.append("select * from fledge.service_schema where name = '");
+	        sql.append(name);
+	      	sql.append("';");
+                const char *query = sql.coalesce();
+                logSQL("findSchemaAndVersionFromDB", query);
+
+                PGresult *res = PQexec(dbConnection, query);
+                delete[] query;
+                if (PQresultStatus(res) == PGRES_TUPLES_OK)
+                {
+                        mapResultSet(res, resultSet);
+                        PQclear(res);
+
+                        return true;
+		}
+		else
+		{
+			char *SQLState = PQresultErrorField(res, PG_DIAG_SQLSTATE);
+                	if (!strcmp(SQLState, "22P02")) // Conversion error
+                 	{
+                        	raiseError("findSchemaAndVersionFromDB", "Unable to convert data to the required type");
+                 	}
+                 	else
+                 	{
+                        	raiseError("findSchemaAndVersionFromDB", PQerrorMessage(dbConnection));
+                 	}
+                 	PQclear(res);
+                 	return false;
+         	}
+	}catch (exception e) {
+                	raiseError("findSchemaAndVersionFromDB", "Internal error: %s", e.what());
+        }
+         
+	return false;
+}
+
+bool Connection::parseDatabaseStorageSchema(int &version, std::string res, 
+		 std::unordered_map<std::string, std::unordered_set<std::string> > &tableColumnMap,
+		 std::unordered_map<std::string, std::unordered_set<std::string> > &tableIndexMap)
+{
+ 	Logger::getLogger()->error( "%s:%d result from table = %s", __FUNCTION__, __LINE__, res.c_str());
+	Document document;
+
+	if (document.Parse(res.c_str()).HasParseError())
+        {
+       		raiseError("checkSchemaUpdateRequired", "Failed to parse JSON payload");
+	        return false;
+        }
+        if (!document.HasMember("rows"))
+        {
+        	Logger::getLogger()->error("schema absent from payload, so exiting ");
+        	return false;
+        }
+        else
+	{
+		Value& rows = document["rows"];
+                if (!rows.IsArray())
+                {
+                	raiseError("checkSchemaUpdateRequired", "The property return must be an array");
+                        return false;
+                }
+                else
+                {
+			version = rows[0]["version"].GetInt();
+
+			Logger::getLogger()->error( "%s:%d version %d", __FUNCTION__, __LINE__,  version);
+
+			Value& def = rows[0]["definition"];
+
+			if (!def.IsObject())
+			{
+				raiseError("checkSchemaUpdateRequired", "The property definition must be an object");
+				return false;
+			}
+
+			Value& tables = def["tables"];
+			if (!tables.IsArray())
+			{
+				raiseError("checkSchemaUpdateRequired", "The property definition must be an array");
```

Comment:
Can you be more descriptive in the error response, "The tables definition should be an array" - it helps the developers of services that expand the database to find the errors more quickly.

Reply:
Fixed